src/tir/schedule/primitive/blockize_tensorize.cc

```diff
@@ -0,0 +1,638 @@
+  // the tensor intrinsic
+
+  // Step 3.1: Map from intrinsic func buffer to description func buffer
+  std::unordered_map<Buffer, Buffer, ObjectPtrHash, ObjectPtrEqual> intrin_buffer_map;
+  RemapTensorIntrinBuffers(intrinsic, &intrin_buffer_map);
+  // Step 3.2: Map form intrinsic func buffer to current AST buffer
+  std::unordered_map<Buffer, Buffer, ObjectPtrHash, ObjectPtrEqual> buffer_map;
+  for (const auto& pair : intrin_buffer_map) {
+    auto it = comparator.rhs_buffer_map_.find(pair.second);
+    ICHECK(it != comparator.rhs_buffer_map_.end());
+    buffer_map[pair.first] = it->second;
+  }
+
+  // Step 4: Create MatchBufferRegion for the params of the implementation function of the tensor
+  // intrin to make them subregions of the buffer in the original IR.
+  std::unordered_map<Buffer, Array<Range>, ObjectPtrHash, ObjectPtrEqual> buffer_region_map;
+  for (const auto& read : impl_block->reads) {
+    buffer_region_map.emplace(read->buffer, read->region);
+  }
+  for (const auto& write : impl_block->writes) {
+    buffer_region_map.emplace(write->buffer, write->region);
+  }
+  Array<MatchBufferRegion> match_buffer_regions;
+  for (size_t i = 0; i < intrinsic->implementation->params.size(); ++i) {
+    const auto& param = intrinsic->implementation->params[i];
+    const auto& buffer = intrinsic->implementation->buffer_map.at(param);
+    const auto& source = buffer_map.at(buffer);
+    Region region = buffer_region_map.at(buffer);
+    auto extra_indices = comparator.buffer_indices_.at(source);
+    std::vector<Range> extra_buffer_ranges;
+    std::transform(extra_indices.begin(), extra_indices.end(),
+                   std::back_inserter(extra_buffer_ranges),
+                   [](const PrimExpr& index) { return Range::FromMinExtent(index, 1); });
+    region.insert(region.begin(), extra_buffer_ranges.begin(), extra_buffer_ranges.end());
+    match_buffer_regions.push_back(MatchBufferRegion(buffer, BufferRegion(source, region)));
+  }
+
+  // Step 5: Replace the subtree in the original IR with the tensor intrin implementation.
+  ObjectPtr<BlockNode> new_block_ptr = make_object<BlockNode>(*block_realize->block.get());
+  new_block_ptr->body = impl_block->body;
+  ICHECK(new_block_ptr->match_buffers.empty());
+  new_block_ptr->match_buffers = std::move(match_buffer_regions);
+  Block new_block(new_block_ptr);
+
+  // Substitution map for the intrin block to get the correct bindings.
+  Map<Var, PrimExpr> bv_map;
+  for (size_t i = 0; i < desc_block->iter_vars.size(); ++i) {
+    auto it = comparator.equal_map_.find(desc_block->iter_vars[i]->var);
+    if (it != comparator.equal_map_.end()) {
+      bv_map.Set(impl_block->iter_vars[i]->var, Downcast<PrimExpr>(it->second));
+    } else {
+      bv_map.Set(impl_block->iter_vars[i]->var, Integer(0));
+    }
+  }
+  new_block = Downcast<Block>(SubstituteInScope(new_block, bv_map));
+  self->Replace(block_sref, new_block, {{block_realize->block, new_block}});
```

Comment:
I know it's hard, but am just curious if we could gather all the mutation logic to one place in the end, so that once there is unsatisfied condition preventing full tensorization, the IR isn't mutated by `Blockize`

Reply:
`Tensorize` now allows to pass in a block instead of a loop. In that case, we don't need to do blockize.

Reply:
I think you didn't get my ideas, but it's not high priority to do in this PR so I'm just curious.

Given we might care about the "exception-safety" of TIR schedule, i.e. if a schedule primitive fails, then we might want the IR to be unchanged. So in our particular case, during tensorization, if `Blockize` works, but the subsequence operation in `Tensorize` fails, then will it leave the IR changed?

Reply:
it's possible, we need to modify `blockize` to have a dry-run version that returns the new block directly and doesn't call `Replace`

Reply:
Thanks for the clarification! Let's focus on landing this PR first and do misc improvement in the future :-)